provider/oci/environ.go

```diff
@@ -223,42 +379,324 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+	args.ImageMetadata = imgCache.ImageMetadata(series, *args.Constraints.VirtType)
+
+	spec, image, err := findInstanceSpec(
+		args.ImageMetadata,
+		types,
+		&instances.InstanceConstraint{
+			Series:      series,
+			Arches:      arches,
+			Constraints: args.Constraints,
+		},
+	)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("agent binaries: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		e.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+	hostname, err := e.namespace.Hostname(args.InstanceConfig.MachineId)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tags := args.InstanceConfig.Tags
+
+	var apiPort int
+	var desiredStatus ociCore.InstanceLifecycleStateEnum
+	// If we are bootstrapping a new controller, we want to wait for the
+	// machine to reach running state before attempting to SSH into it,
+	// to configure the controller.
+	// If the machine that is spawning is not a controller, then userdata
+	// will take care of it's initial setup, and waiting for a running
+	// status is not necessary
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+		desiredStatus = ociCore.InstanceLifecycleStateRunning
+	} else {
+		desiredStatus = ociCore.InstanceLifecycleStateProvisioning
+	}
+
+	cloudcfg, err := e.getCloudInitConfig(series, apiPort)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	logger.Debugf("Composing userdata")
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OCIRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	var rootDiskSize int
+	if args.Constraints.RootDisk != nil {
+		// apparently, the RootDisk constraint is expressed in KB
+		rootDiskSize = int(*args.Constraints.RootDisk) / 1024
+		if int(*args.Constraints.RootDisk) < MinVolumeSizeMB {
+			logger.Warningf(
+				"selected disk size is too small (%d). Setting root disk size to minimum volume size (%d)",
+				int(*args.Constraints.RootDisk), MinVolumeSizeMB)
+			rootDiskSize = MinVolumeSizeMB / 1024
+		} else if int(*args.Constraints.RootDisk) > MaxVolumeSizeMB {
+			logger.Warningf(
+				"selected disk size is too large (%d). Setting root disk size to maximum volume size (%d)",
+				int(*args.Constraints.RootDisk), MaxVolumeSizeMB)
+			rootDiskSize = MaxVolumeSizeMB / 1024
+		}
+	} else {
+		rootDiskSize = MinVolumeSizeMB / 1024
+	}
+
+	assignPublicIp := true
+	bootSource := ociCore.InstanceSourceViaImageDetails{
+		ImageId:             &image,
+		BootVolumeSizeInGBs: &rootDiskSize,
```

Comment:
It would appear that it is just your comments that are wrong.
RootDisk as a *constraint* is in MB, Min and Max VolumeSizeMB are in MB, but the BootVolumeSizeInGBs is in GB.

I would go back and rename the variable
rootDiskSizeGB (or InGB), and then I think your conversions are, in fact, correct. I might drop the comments, as they were the bit that actually confused me.